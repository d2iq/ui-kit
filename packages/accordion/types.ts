export type HeadingLevel = 2 | 3 | 4 | 5 | 6;
export type AccordionTitleAppearances = "default" | "danger";
