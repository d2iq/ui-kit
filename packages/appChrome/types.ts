import { BreakpointConfig } from "../shared/styles/breakpoints";
import { SpaceSizes } from "../shared/styles/styleUtils/modifiers/modifierUtils";
import { LinkProps } from "../link/types";
import { IconSize } from "../shared/types/iconSize";

type BgColor = React.CSSProperties["backgroundColor"];
type PaddingHoriz = SpaceSizes;
type PaddingVert = SpaceSizes;

export interface AppChromeTheme {
  headerBackgroundColor?: BgColor;
  headerPaddingHor?: PaddingHoriz;
  headerPaddingVert?: PaddingVert;
  iconWidth?: IconSize;
  itemActiveBackgroundColor?: BgColor;
  itemHoverBackgroundColor?: BgColor;
  sidebarBackgroundColor?: BgColor;
  sidebarHeaderPaddingHor?: PaddingHoriz;
  sidebarHeaderPaddingVert?: PaddingVert;
  sidebarWidth?: BreakpointConfig<React.CSSProperties["width"]>;
  sidebarItemPaddingHor?: PaddingHoriz;
  sidebarItemPaddingVert?: PaddingVert;
}

export interface SidebarNavItemProps extends LinkProps {
  isActive?: boolean;
  onClick?: (event?: React.SyntheticEvent<HTMLElement>) => void;
  disabled?: boolean;
}
