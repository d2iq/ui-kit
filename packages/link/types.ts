import React from "react";

export interface LinkProps {
  /**
   * A url the user will be navigated to when clicking the link. This also changes the tag to `<a>`.
   */
  url?: string;
  /**
   * If the `url` prop is set, this will open that link in a new tab.
   */
  openInNewTab?: boolean;
  children?: React.ReactNode | React.ReactNode[];
}

export type ExpandedLinkProps = LinkProps &
  Omit<React.HTMLProps<HTMLAnchorElement>, "ref">;

export type LinkComponent = React.ComponentType<
  LinkProps & React.HTMLProps<HTMLAnchorElement>
>;
