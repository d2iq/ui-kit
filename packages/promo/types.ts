export interface PromoProps {
  /** The primary banner content */
  bodyContent: React.ReactNode;
  /** A function that gets called when the banner is dismissed */
  dismissHandler?: () => void;
  /** A consise summary of the banner's message */
  headingText: string;
  /** Whether the content is on a dark background. Changes text color to be readable. */
  isDarkBackground?: boolean;
  /** A link to the graphic's image file. The graphic (typically an illustration) should be relevant to the banner's message. Ideally the graphic should not just be decorative, but it reinforces the banner's overall message. */
  graphicSrc?: string;
  /** Whether a user has opted not to see the banner */
  optOutBanner?: boolean;
  /** A function that gets called when the user checks the "Do not show again" checkbox */
  optOutHandler?: (e: React.SyntheticEvent<HTMLInputElement>) => void;
  /** The most relevant action a user can take in response to the banner's message */
  primaryAction?: React.ReactNode;
  /** An additional action a user can take in response to the banner's message. This can also just be a link to more information relevant to the banner's message. */
  secondaryAction?: React.ReactNode;
  /** The Cypress selector */
  ["data-cy"]?: string;
}

export type GradientStyle = "lightBlue" | "purple" | "oceanBlue";
export type GradientColors = [string, string, string?];
export type PromoBackgroundColor = "themeBgSecondary" | "purpleLighten5";
