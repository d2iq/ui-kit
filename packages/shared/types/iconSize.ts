export type IconSize = "xxs" | "xs" | "s" | "m" | "l" | "xl" | "xxl";
