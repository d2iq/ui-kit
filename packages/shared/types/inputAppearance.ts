export enum InputAppearance {
  Standard = "standard",
  Error = "error",
  Success = "success"
}
