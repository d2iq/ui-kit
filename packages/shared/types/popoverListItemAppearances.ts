export enum PopoverListItemAppearances {
  DANGER = "danger"
}
