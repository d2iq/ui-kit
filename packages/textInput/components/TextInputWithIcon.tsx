import { cx } from "@emotion/css";
import * as React from "react";

import { TextInput, TextInputProps } from "./TextInput";

import {
  getInputAppearanceStyle,
  inputContainer,
  getIconAppearanceStyle,
  inputIconWrapper
} from "../../shared/styles/formStyles";
import { flex, flexItem, flush, padding } from "../../shared/styles/styleUtils";
import FormFieldWrapper from "../../shared/components/FormFieldWrapper";
import { InputAppearance } from "../../shared/types/inputAppearance";
import { IconShapes } from "../../icon/components/Icon";
import IconPropAdapter from "../../icon/components/IconPropAdapter";

export interface TextInputWithIconProps extends TextInputProps {
  /**
   * icon to display at the start of this TextInput.
   */
  iconStart?: IconShapes | React.ReactElement<HTMLElement>;
  /**
   * icon to display at the end of this TextInput.
   */
  iconEnd?: IconShapes | React.ReactElement<HTMLElement>;
}

export interface TextInputWithIconState {
  hasFocus?: boolean;
}

export class TextInputWithIcon<
  P extends TextInputWithIconProps
> extends TextInput<P> {
  public static defaultProps = {
    type: "text",
    appearance: InputAppearance.Standard,
    showInputLabel: true
  };

  state = {
    hasFocus: false
  };

  protected getInputAppearance(): string {
    if (this.props.disabled) {
      return "disabled";
    }
    if (this.state.hasFocus) {
      return `${this.props.appearance}-focus`;
    }
    return this.props.appearance;
  }

  protected getInputElementProps() {
    const baseProps = super.getInputElementProps();
    const {
      iconStart,
      iconEnd,
      ...inputProps
    } = baseProps as TextInputWithIconProps;
    inputProps.onFocus = this.inputOnFocus;
    inputProps.onBlur = this.inputOnBlur;
    return inputProps;
  }

  protected getIconStartContent() {
    if (!this.props.iconStart) {
      return null;
    }
    return (
      <span
        className={cx(
          padding("right", "xs"),
          flex({ align: "center", justify: "center" }),
          flexItem("shrink"),
          getIconAppearanceStyle(this.getInputAppearance()),
          inputIconWrapper
        )}
      >
        <IconPropAdapter
          icon={this.props.iconStart!}
          size="xs"
          color="inherit"
        />
      </span>
    );
  }

  protected getIconEndContent() {
    if (!this.props.iconEnd) {
      return null;
    }
    return (
      <span
        className={cx(
          flex({ align: "center", justify: "center" }),
          flexItem("shrink"),
          flush("left"),
          getIconAppearanceStyle(this.getInputAppearance()),
          inputIconWrapper
        )}
      >
        <IconPropAdapter icon={this.props.iconEnd!} size="xs" color="inherit" />
      </span>
    );
  }

  protected getInputContent() {
    const inputAppearance = this.getInputAppearance();
    return (
      <FormFieldWrapper
        // TODO: figure out how to get rid of this non-null assertion
        // If we stop generating an `id` prop in the TextInput component,
        // it would be possible for `this.props.id` to be undefined
        id={this.props.id!}
        errors={this.props.errors}
        hintContent={this.props.hintContent}
      >
        {({ getValidationErrors, getHintContent, isValid, describedByIds }) => (
          <div>
            <div
              className={cx(
                flex(),
                padding("left", "s"),
                padding("right", "s"),
                inputContainer,
                getInputAppearanceStyle(inputAppearance)
              )}
            >
              {this.getIconStartContent()}
              {this.getInputElement(
                [flexItem("grow"), padding("all", "none")],
                isValid,
                describedByIds
              )}
              {this.getIconEndContent()}
            </div>
            {getHintContent}
            {getValidationErrors}
          </div>
        )}
      </FormFieldWrapper>
    );
  }

  protected inputOnFocus = e => {
    this.setState({ hasFocus: true });

    if (this.props.onFocus) {
      this.props.onFocus(e);
    }
  };

  protected inputOnBlur = e => {
    this.setState({ hasFocus: false });

    if (this.props.onBlur) {
      this.props.onBlur(e);
    }
  };
}

export default TextInputWithIcon;
