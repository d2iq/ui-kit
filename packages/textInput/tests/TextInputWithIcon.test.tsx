import React from "react";
import serializer from "@emotion/jest";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import TextInputWithIcon from "../components/TextInputWithIcon";
import { InputAppearance } from "../../shared/types/inputAppearance";
import { SystemIcons } from "../../icons/dist/system-icons-enum";

expect.addSnapshotSerializer(serializer);

describe("TextInputWithIcon", () => {
  it("should render all appearances with iconStart", () => {
    Object.keys(InputAppearance).forEach(appearance => {
      const component = shallow(
        <TextInputWithIcon
          id={`test.icon_start_input.${InputAppearance[appearance]}`}
          inputLabel={InputAppearance[appearance]}
          appearance={InputAppearance[appearance]}
          iconStart={SystemIcons.TriangleDown}
        />
      );
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  it("should render all appearances with iconEnd", () => {
    Object.keys(InputAppearance).forEach(appearance => {
      const component = shallow(
        <TextInputWithIcon
          id={`test.icon_end_input.${InputAppearance[appearance]}`}
          inputLabel={InputAppearance[appearance]}
          appearance={InputAppearance[appearance]}
          iconEnd={SystemIcons.Close}
        />
      );
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  it("should render all appearances with twoIcons", () => {
    Object.keys(InputAppearance).forEach(appearance => {
      const component = shallow(
        <TextInputWithIcon
          id={`test.two_icon_input.${InputAppearance[appearance]}`}
          inputLabel={InputAppearance[appearance]}
          appearance={InputAppearance[appearance]}
          iconStart={SystemIcons.TriangleDown}
          iconEnd={SystemIcons.Close}
        />
      );
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  it("should call onFocus when input gains focus", () => {
    const focusFn = jest.fn();
    const component = mount(
      <TextInputWithIcon
        id="test.input"
        inputLabel="Test Focus"
        onFocus={focusFn}
        iconStart={SystemIcons.TriangleDown}
      />
    );
    expect(focusFn).not.toHaveBeenCalled();
    component.find("input").simulate("focus");
    expect(focusFn).toHaveBeenCalled();
  });

  it("should call onBlur when input loses focus", () => {
    const blurFn = jest.fn();
    const component = mount(
      <TextInputWithIcon
        id="test.input"
        inputLabel="Test Blur"
        onBlur={blurFn}
        iconStart={SystemIcons.TriangleDown}
      />
    );
    expect(blurFn).not.toHaveBeenCalled();
    component.find("input").simulate("focus");
    expect(blurFn).not.toHaveBeenCalled();
    component.find("input").simulate("blur");
    expect(blurFn).toHaveBeenCalled();
  });
});
