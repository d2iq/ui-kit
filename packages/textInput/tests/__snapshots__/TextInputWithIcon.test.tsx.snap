// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputWithIcon should render all appearances with iconEnd 1`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.standard"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_end_input.standard"
  >
    standard
  </label>
  <FormFieldWrapper
    id="test.icon_end_input.standard"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with iconEnd 2`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
  color: var(--themeError, #EB293A);
  fill: var(--themeError, #EB293A);
}

<div
  data-cy="textInput textInput.error"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_end_input.error"
  >
    error
  </label>
  <FormFieldWrapper
    id="test.icon_end_input.error"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with iconEnd 3`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.success"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_end_input.success"
  >
    success
  </label>
  <FormFieldWrapper
    id="test.icon_end_input.success"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with iconStart 1`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.standard"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_start_input.standard"
  >
    standard
  </label>
  <FormFieldWrapper
    id="test.icon_start_input.standard"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with iconStart 2`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
  color: var(--themeError, #EB293A);
  fill: var(--themeError, #EB293A);
}

<div
  data-cy="textInput textInput.error"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_start_input.error"
  >
    error
  </label>
  <FormFieldWrapper
    id="test.icon_start_input.error"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with iconStart 3`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.success"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.icon_start_input.success"
  >
    success
  </label>
  <FormFieldWrapper
    id="test.icon_start_input.success"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with twoIcons 1`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.standard"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.two_icon_input.standard"
  >
    standard
  </label>
  <FormFieldWrapper
    id="test.two_icon_input.standard"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with twoIcons 2`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
  color: var(--themeError, #EB293A);
  fill: var(--themeError, #EB293A);
}

<div
  data-cy="textInput textInput.error"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.two_icon_input.error"
  >
    error
  </label>
  <FormFieldWrapper
    id="test.two_icon_input.error"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;

exports[`TextInputWithIcon should render all appearances with twoIcons 3`] = `
.emotion-0 {
  display: block;
  padding-top: 0;
  margin-top: 0;
  margin-bottom: 4px;
  font-weight: 500;
}

<div
  data-cy="textInput textInput.success"
>
  <label
    className="emotion-0"
    data-cy="textInput-label"
    htmlFor="test.two_icon_input.success"
  >
    success
  </label>
  <FormFieldWrapper
    id="test.two_icon_input.success"
  >
    <Component />
  </FormFieldWrapper>
</div>
`;
